File: conformance/runner/go/schema_walker.go
package main

// Pure-Go port of conformance/runner/typescript/schema-validator.ts.
//
// Walks parsed JSON against the OpenAPI response schema, surfacing:
//   - MissingRequired: slash-separated paths for required fields absent
//     from the body (e.g. "owner/id")
//   - ExtrasSeen: dotted paths for fields present on the wire but not
//     declared (e.g. "unreads[].new_field")
//
// Conventions match the TS walker (and the Python/Ruby ports under
// conformance/runner/{python,ruby}/) so cross-language extras parity diffs
// (PR 4 §Verification) don't false-fire:
//   - Required-walk object paths use "/" (e.g. "owner/id"); extras-walk
//     object paths use "." (e.g. "owner.new_field"). The two streams use
//     distinct separators so they're visually distinguishable in tooling.
//   - Required walk uses "[i]" element segments; extras walk uses "[]"
//     to dedupe item-level extras across an array
//   - "$ref" chains resolve until a non-ref schema or a cycle. Both
//     "#/components/schemas/X" and "openapi.json#/components/schemas/X"
//     are accepted.
//   - additionalProperties:false is intentionally ignored — extras are
//     reported but do not fail validation (forward-compat).
//   - Recursion depth bound 12 as a cycle guard.
//
// No new dependencies: hand-rolled keeps semantics identical to the TS
// walker. After json.Unmarshal into any, JSON objects become
// map[string]any and arrays become []any; both branches dispatch on that.

import (
	"encoding/json"
	"os"
	"regexp"
	"sort"
)

const maxDepth = 12

var refRE = regexp.MustCompile(`^(?:openapi\.json)?#/components/schemas/(.+)$`)

type SchemaWalker struct {
	doc map[string]any
}

func NewSchemaWalker(openapiPath string) (*SchemaWalker, error) {
	raw, err := os.ReadFile(openapiPath)
	if err != nil {
		return nil, err
	}
	var doc map[string]any
	if err := json.Unmarshal(raw, &doc); err != nil {
		return nil, err
	}
	return &SchemaWalker{doc: doc}, nil
}

// FindResponseSchema returns the response schema for operationID, or nil
// when none is found. Match TS preference order: 200, then any 2xx, then
// "default".
func (w *SchemaWalker) FindResponseSchema(operationID string) map[string]any {
	paths, _ := w.doc["paths"].(map[string]any)
	for _, pathName := range sortedKeys(paths) {
		pathItem, _ := paths[pathName].(map[string]any)
		for _, method := range sortedKeys(pathItem) {
			op, ok := pathItem[method].(map[string]any)
			if !ok {
				continue
			}
			if id, _ := op["operationId"].(string); id != operationID {
				continue
			}
			responses, _ := op["responses"].(map[string]any)
			for _, code := range []string{"200", "201", "202", "203", "204", "default"} {
				if s := schemaFor(responses[code]); s != nil {
					return s
				}
			}
			for _, code := range sortedKeys(responses) {
				if len(code) == 3 && code[0] == '2' && isDigit(code[1]) && isDigit(code[2]) {
					if s := schemaFor(responses[code]); s != nil {
						return s
					}
				}
			}
		}
	}
	return nil
}

// MissingRequired returns slash-separated path strings for required fields
// absent from body (e.g. "owner/id").
func (w *SchemaWalker) MissingRequired(body any, schema map[string]any) []string {
	return w.walkRequired("", body, schema, 0)
}

// ExtrasSeen returns dotted-path strings for fields present on the wire
// but not declared in the schema. Recurses through known properties so
// item-level extras on lists surface (e.g. "unreads[].new_field").
func (w *SchemaWalker) ExtrasSeen(body any, schema map[string]any) []string {
	return w.walkExtras("", body, schema, 0)
}

func (w *SchemaWalker) walkRequired(prefix string, body, schema any, depth int) []string {
	if depth > maxDepth || body == nil {
		return nil
	}
	resolved := w.resolveRef(schema)
	s, ok := resolved.(map[string]any)
	if !ok {
		return nil
	}

	if arr, isArr := body.([]any); isArr {
		typ, _ := s["type"].(string)
		if typ != "array" || s["items"] == nil {
			return nil
		}
		var missing []string
		for i, item := range arr {
			child := joinIndex(prefix, i)
			missing = append(missing, w.walkRequired(child, item, s["items"], depth+1)...)
		}
		return missing
	}

	obj, isObj := body.(map[string]any)
	if !isObj {
		return nil
	}
	if typ, present := s["type"].(string); present && typ != "object" {
		return nil
	}

	props, _ := s["properties"].(map[string]any)
	required, _ := s["required"].([]any)
	var missing []string
	// Required-field paths use `/` as the separator (walkExtras uses `.`)
	// so the two streams are visually distinct in tooling and consistent
	// across Ruby/Python/Go/Kotlin walkers.
	for _, r := range required {
		name, _ := r.(string)
		if _, ok := obj[name]; !ok {
			missing = append(missing, joinSlash(prefix, name))
		}
	}
	// Recurse into present known props. Sort keys: cross-language
	// comparable output (TS preserves JSON parse order; Go map iteration
	// is randomized, so sorted keys keep diffs stable).
	for _, name := range sortedKeys(props) {
		if value, ok := obj[name]; ok {
			missing = append(missing, w.walkRequired(joinSlash(prefix, name), value, props[name], depth+1)...)
		}
	}
	return missing
}

func (w *SchemaWalker) walkExtras(prefix string, body, schema any, depth int) []string {
	if depth > maxDepth || body == nil {
		return nil
	}
	resolved := w.resolveRef(schema)
	s, ok := resolved.(map[string]any)
	if !ok {
		return nil
	}

	if arr, isArr := body.([]any); isArr {
		typ, _ := s["type"].(string)
		if typ != "array" || s["items"] == nil {
			return nil
		}
		// Per-array dedup mirrors TS collectExtras's `new Set` for arrays.
		seen := map[string]struct{}{}
		var out []string
		child := joinBrackets(prefix)
		for _, item := range arr {
			for _, e := range w.walkExtras(child, item, s["items"], depth+1) {
				if _, dup := seen[e]; !dup {
					seen[e] = struct{}{}
					out = append(out, e)
				}
			}
		}
		return out
	}

	obj, isObj := body.(map[string]any)
	if !isObj {
		return nil
	}
	if typ, present := s["type"].(string); present && typ != "object" {
		return nil
	}

	props, _ := s["properties"].(map[string]any)
	var extras []string
	// Sort body keys: stable cross-language output. TS preserves JSON parse
	// order, which equals insertion order; Go map iteration is randomized,
	// so we sort. PR 4 cross-lang extras parity diff compares deduped sets
	// anyway, so this only affects the per-page ordering inside the snapshot.
	for _, key := range sortedKeys(obj) {
		fieldPath := joinDot(prefix, key)
		if _, known := props[key]; known {
			extras = append(extras, w.walkExtras(fieldPath, obj[key], props[key], depth+1)...)
		} else {
			extras = append(extras, fieldPath)
		}
	}
	return extras
}

// resolveRef follows $ref chains until a non-ref schema or a cycle.
// Accepts "#/components/schemas/X" and "openapi.json#/components/schemas/X".
func (w *SchemaWalker) resolveRef(schema any) any {
	seen := map[string]struct{}{}
	current := schema
	for {
		m, ok := current.(map[string]any)
		if !ok {
			return current
		}
		ref, ok := m["$ref"].(string)
		if !ok {
			return current
		}
		if _, dup := seen[ref]; dup {
			return current
		}
		seen[ref] = struct{}{}
		match := refRE.FindStringSubmatch(ref)
		if match == nil {
			return current
		}
		components, _ := w.doc["components"].(map[string]any)
		schemas, _ := components["schemas"].(map[string]any)
		next, ok := schemas[match[1]]
		if !ok {
			return current
		}
		current = next
	}
}

func schemaFor(response any) map[string]any {
	resp, ok := response.(map[string]any)
	if !ok {
		return nil
	}
	content, _ := resp["content"].(map[string]any)
	appJSON, _ := content["application/json"].(map[string]any)
	schema, _ := appJSON["schema"].(map[string]any)
	return schema
}

func sortedKeys(m map[string]any) []string {
	if m == nil {
		return nil
	}
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

func isDigit(b byte) bool { return b >= '0' && b <= '9' }

func joinDot(prefix, name string) string {
	if prefix == "" {
		return name
	}
	return prefix + "." + name
}

func joinSlash(prefix, name string) string {
	if prefix == "" {
		return name
	}
	return prefix + "/" + name
}

func joinIndex(prefix string, i int) string {
	suffix := "[" + itoa(i) + "]"
	if prefix == "" {
		return suffix
	}
	return prefix + suffix
}

func joinBrackets(prefix string) string {
	if prefix == "" {
		return "[]"
	}
	return prefix + "[]"
}

// itoa avoids importing strconv just for one use.
func itoa(i int) string {
	if i == 0 {
		return "0"
	}
	neg := false
	if i < 0 {
		neg = true
		i = -i
	}
	var buf [20]byte
	pos := len(buf)
	for i > 0 {
		pos--
		buf[pos] = byte('0' + i%10)
		i /= 10
	}
	if neg {
		pos--
		buf[pos] = '-'
	}
	return string(buf[pos:])
}

File: kotlin/conformance/src/main/kotlin/com/basecamp/sdk/conformance/SchemaWalker.kt
package com.basecamp.sdk.conformance

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.File

/**
 * Pure-Kotlin port of `conformance/runner/typescript/schema-validator.ts`.
 *
 * Walks parsed JSON against the OpenAPI response schema, surfacing:
 *   - missing required-field paths (slash-separated, e.g. "owner/id")
 *   - extras-seen field paths (dot-separated for objects, "[]" for arrays,
 *     e.g. "unreads[].new_field")
 *
 * Conventions match the TS / Ruby / Python / Go ports in
 * `conformance/runner/{ruby,python,go}/` so cross-language extras parity diffs
 * (PR 4 §Verification) don't false-fire:
 *   - `additionalProperties: false` is intentionally ignored — extras are
 *     reported but do not fail validation (forward-compat).
 *   - `$ref` chains resolve until a non-ref schema or a cycle. Both
 *     `#/components/schemas/X` and `openapi.json#/components/schemas/X` are
 *     accepted.
 *   - Recursion depth bound 12 as a cycle guard.
 *   - Required walk uses `[i]` element segments; extras walk uses `[]` to
 *     dedupe item-level extras across an array.
 *   - A field whose value is `JsonNull` counts as PRESENT (matches TS
 *     `name in body` semantics) — `nullable` separately governs whether null
 *     is acceptable, but presence-checking only asks "is the key in the
 *     object?".
 *
 * No new dependencies — kotlinx.serialization.json is already on the
 * conformance build classpath via the mock runner.
 */
class SchemaWalker(openapiPath: String) {
    private val doc: JsonObject = Json.parseToJsonElement(File(openapiPath).readText()).jsonObject
    private val schemas: JsonObject = doc["components"]?.jsonObject?.get("schemas")?.jsonObject
        ?: JsonObject(emptyMap())

    /**
     * Returns null when no response schema is found (do not throw).
     * Preference order: 200, then any 2xx, then "default" — matches TS.
     */
    fun findResponseSchema(operationId: String): JsonObject? {
        val paths = doc["paths"]?.jsonObject ?: return null
        for ((_, pathItemEl) in paths) {
            val pathItem = pathItemEl as? JsonObject ?: continue
            for ((_, opEl) in pathItem) {
                val op = opEl as? JsonObject ?: continue
                val opId = op["operationId"]?.jsonPrimitive?.contentOrNull
                if (opId != operationId) continue
                val responses = op["responses"]?.jsonObject ?: continue
                for (code in PREFERRED_CODES) {
                    schemaFor(responses[code])?.let { return it }
                }
                for ((code, response) in responses) {
                    if (code.length == 3 && code[0] == '2' && code[1].isAsciiDigit() && code[2].isAsciiDigit()) {
                        schemaFor(response)?.let { return it }
                    }
                }
                schemaFor(responses["default"])?.let { return it }
            }
        }
        return null
    }

    /** Slash-separated paths for required fields absent from body. */
    fun missingRequired(body: JsonElement, schema: JsonObject): List<String> {
        val out = mutableListOf<String>()
        walkRequired("", body, schema, 0, out)
        return out
    }

    /** Dotted-path strings for fields present on wire but not declared in schema. */
    fun extrasSeen(body: JsonElement, schema: JsonObject): List<String> {
        val out = mutableListOf<String>()
        walkExtras("", body, schema, 0, out)
        return out
    }

    private fun walkRequired(prefix: String, body: JsonElement, schema: JsonElement, depth: Int, out: MutableList<String>) {
        if (depth > MAX_DEPTH || body is JsonNull) return
        val s = resolveRef(schema) as? JsonObject ?: return

        if (body is JsonArray) {
            if (s["type"]?.jsonPrimitive?.contentOrNull != "array") return
            val items = s["items"] ?: return
            for ((i, item) in body.withIndex()) {
                val child = if (prefix.isEmpty()) "[$i]" else "$prefix[$i]"
                walkRequired(child, item, items, depth + 1, out)
            }
            return
        }
        if (body !is JsonObject) return
        val type = s["type"]?.jsonPrimitive?.contentOrNull
        if (type != null && type != "object") return

        val props = s["properties"]?.jsonObject ?: JsonObject(emptyMap())
        // Emit prefix/name for any required key absent from the body. JsonNull
        // values count as PRESENT (TS `name in body`).
        val required = s["required"] as? JsonArray
        if (required != null) {
            for (req in required) {
                val name = req.jsonPrimitive.contentOrNull ?: continue
                if (!body.containsKey(name)) {
                    out += if (prefix.isEmpty()) name else "$prefix/$name"
                }
            }
        }
        for ((name, sub) in props) {
            val value = body[name] ?: continue
            val child = if (prefix.isEmpty()) name else "$prefix/$name"
            walkRequired(child, value, sub, depth + 1, out)
        }
    }

    private fun walkExtras(prefix: String, body: JsonElement, schema: JsonElement, depth: Int, out: MutableList<String>) {
        if (depth > MAX_DEPTH || body is JsonNull) return
        val s = resolveRef(schema) as? JsonObject ?: return

        if (body is JsonArray) {
            if (s["type"]?.jsonPrimitive?.contentOrNull != "array") return
            val items = s["items"] ?: return
            val child = if (prefix.isEmpty()) "[]" else "$prefix[]"
            // Per-array dedup mirrors TS collectExtras's `new Set` for arrays.
            val seen = mutableSetOf<String>()
            for (item in body) {
                val tmp = mutableListOf<String>()
                walkExtras(child, item, items, depth + 1, tmp)
                for (e in tmp) if (seen.add(e)) out += e
            }
            return
        }
        if (body !is JsonObject) return
        val type = s["type"]?.jsonPrimitive?.contentOrNull
        if (type != null && type != "object") return

        val props = s["properties"]?.jsonObject ?: JsonObject(emptyMap())
        for ((key, value) in body) {
            val fieldPath = if (prefix.isEmpty()) key else "$prefix.$key"
            val sub = props[key]
            if (sub == null) {
                out += fieldPath
            } else {
                walkExtras(fieldPath, value, sub, depth + 1, out)
            }
        }
    }

    /** Follow $ref chains until a non-ref schema or a cycle. */
    private fun resolveRef(schema: JsonElement): JsonElement {
        val seen = mutableSetOf<String>()
        var current: JsonElement = schema
        while (current is JsonObject) {
            val ref = current["\$ref"]?.jsonPrimitive?.contentOrNull ?: return current
            if (!seen.add(ref)) return current
            val match = REF_REGEX.matchEntire(ref) ?: return current
            current = schemas[match.groupValues[1]] ?: return current
        }
        return current
    }

    private fun schemaFor(response: JsonElement?): JsonObject? {
        val resp = response as? JsonObject ?: return null
        val content = resp["content"] as? JsonObject ?: return null
        val appJson = content["application/json"] as? JsonObject ?: return null
        return appJson["schema"] as? JsonObject
    }

    private fun Char.isAsciiDigit(): Boolean = this in '0'..'9'

    companion object {
        private const val MAX_DEPTH = 12
        private val PREFERRED_CODES = listOf("200", "201", "202", "203", "204")
        private val REF_REGEX = Regex("^(?:openapi\\.json)?#/components/schemas/(.+)$")
    }
}
